cpp/unreal_plugins/SimulationController/Source/SimulationController/WheeledVehicleAgent.cpp

```diff
@@ -0,0 +1,310 @@
+        SP_ASSERT(imu_sensor_);
+    }
+
+    if (Std::contains(observation_components, "sonar")) {
+        sonar_sensor_ = std::make_unique<SonarSensor>(wheeled_vehicle_pawn_->sonar_component_);
+        SP_ASSERT(sonar_sensor_);
+    }
+}
+
+WheeledVehicleAgent::~WheeledVehicleAgent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "sonar")) {
+        SP_ASSERT(sonar_sensor_);
+        sonar_sensor_ = nullptr;
+    }
+
+    if (Std::contains(observation_components, "imu")) {
+        SP_ASSERT(imu_sensor_);
+        imu_sensor_ = nullptr;
+    }
+
+    if (Std::contains(observation_components, "camera")) {
+        SP_ASSERT(camera_sensor_);
+        camera_sensor_ = nullptr;
+    }
+
+    SP_ASSERT(wheeled_vehicle_pawn_);
+    wheeled_vehicle_pawn_->Destroy();
+    wheeled_vehicle_pawn_ = nullptr;
+}
+
+void WheeledVehicleAgent::findObjectReferences(UWorld* world) {}
+
+void WheeledVehicleAgent::cleanUpObjectReferences() {}
+
+std::map<std::string, ArrayDesc> WheeledVehicleAgent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.ACTION_COMPONENTS");
+
+    if (Std::contains(action_components, "apply_wheel_torques")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();;
+        array_desc.high_ = std::numeric_limits<float>::max();;
+        array_desc.shape_ = { 4 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["apply_wheel_torques"] = std::move(array_desc);
+    }
+
+    if (Std::contains(action_components, "set_position_xyz_centimeters")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.shape_ = { 3 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["set_position_xyz_centimeters"] = std::move(array_desc);
+    }
+
+    if (Std::contains(action_components, "set_orientation_pyr_radians")) {
+        ArrayDesc array_desc;
+        array_desc.low_ = std::numeric_limits<float>::lowest();
+        array_desc.high_ = std::numeric_limits<float>::max();
+        array_desc.shape_ = { 3 };
+        array_desc.datatype_ = DataType::Float64;
+        action_space["set_orientation_pyr_radians"] = std::move(array_desc);
+    }
+
+    return action_space;
+}
+
+std::map<std::string, ArrayDesc> WheeledVehicleAgent::getObservationSpace() const
+{
+    std::map<std::string, ArrayDesc> observation_space;
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.WHEELED_VEHICLE_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "state_data")) {
```

Comment:
Factor the observations in a cleaner way: `imu`, `position`, `rotation`, `wheel_encoders`

Reply:
Agreed, updated this.